src/resolve/resolved-dns-stream.c

```diff
@@ -441,6 +439,45 @@ static int on_stream_io(sd_event_source *es, int fd, uint32_t revents, void *use
                         log_warning_errno(errno, "Couldn't restart TCP connection timeout, ignoring: %m");
         }
 
+        return 1;
+}
+
+
+static int on_stream_io(sd_event_source *es, int fd, uint32_t revents, void *userdata) {
+        _cleanup_(dns_stream_unrefp) DnsStream *s = dns_stream_ref(userdata); /* Protect stream while we process it */
+        int r;
+
+        assert(s);
+
+        r = on_stream_io_impl(s, revents);
+        if (r <= 0)
+                return r;
```

Comment:
Could you add a brief comment at the beginning of `on_stream_io_impl()` to mention that it returns negative errno on critical error, 0 when recoverable error or the stream handling is finished. and 1 for possibly remaining stream exists.